Changes in module 'nf-core/sratools/fasterqdump'
--- modules/nf-core/sratools/fasterqdump/main.nf
+++ modules/nf-core/sratools/fasterqdump/main.nf
@@ -43,7 +43,9 @@
         ${key_file} \\
         ${sra}
 
-    $exclude_third
+    if [ -f "${meta.id}_2.fastq" ]; then
+        mv $outfile $prefix || echo 'No third file'
+    fi
 
     pigz \\
         $args2 \\
@@ -86,7 +88,9 @@
         ${key_file} \\
         ${sra}"
 
-    $exclude_third
+    if [ -f "${meta.id}_2.fastq" ]; then
+        mv $outfile $prefix || echo 'No third file'
+    fi
 
     pigz \\
         $args2 \\

************************************************************
